net/http/balancer.go

```diff
@@ -131,41 +136,39 @@ func (b *balancer) countOfBackends(service string) int {
 	return 0
 }
 
-func (b *balancer) nextRoundRobin(service string) (*backend, error) {
-	b.mutex.RLock()
-	defer b.mutex.RUnlock()
-	if upstream, found := b.upstreams[service]; found {
-		upstream.index = (upstream.index + 1) % len(upstream.backends)
-		return upstream.backends[upstream.index], nil
+func (b *balancer) nextRoundRobin(service string, maxRequestsByBackend int) (*backend, error) {
+	upstream := b.getUpstreamByServiceName(service)
+	if upstream != nil {
+		return upstream.nextBackend(maxRequestsByBackend), nil
 	}
 	return nil, fmt.Errorf("Service '%s' not found", service)
 }
 
-func (b *balancer) nextWeight(service string) (*backend, error) {
+func (b *balancer) nextWeight(service string, maxRequestsByBackend int) (*backend, error) {
+	upstream := b.getUpstreamByServiceName(service)
+	if upstream == nil {
+		return nil, fmt.Errorf("Service '%s' not found", service)
+	}
+
+	if backend := upstream.nextWeightBackend(maxRequestsByBackend); backend != nil {
+		return backend, nil
+	}
+
+	return nil, fmt.Errorf("Service backend of '%s' not found", service)
+}
+
+func (b *balancer) next(service string, maxRequestsByBackend int) (*backend, error) {
+	if b.strategy == WeightStrategy {
+		return b.nextWeight(service, maxRequestsByBackend)
+	}
+	return b.nextRoundRobin(service, maxRequestsByBackend)
+}
+
+func (b *balancer) getUpstreamByServiceName(service string) *upstream {
 	b.mutex.RLock()
 	defer b.mutex.RUnlock()
-	if upstream, found := b.upstreams[service]; found {
-		for {
-			upstream.index = (upstream.index + 1) % len(upstream.backends)
-			if upstream.index == 0 {
-				upstream.currentWeight = upstream.currentWeight - upstream.gcd
-				if upstream.currentWeight <= 0 {
-					upstream.currentWeight = upstream.maxWeight
-					if upstream.currentWeight == 0 {
-						return upstream.backends[upstream.index], nil
-					}
-				}
-			}
-			if backend := upstream.backends[upstream.index]; backend.weight >= upstream.currentWeight {
-				if backend.skipCounter != 0 {
-					backend.skipCounter--
-					continue
-				}
-				return backend, nil
-			}
-		}
-	}
-	return nil, fmt.Errorf("Service '%s' not found", service)
+	ups, _ := b.upstreams[service]
```

Comment:
I think it would be better for performance to use `sync.Map` instead of `sync.Mutex` here.

Reply:
Or even better if you will create a storage will simple map and will replace it (on each updated) with `atomic.StorePointer`.

Reply:
Found a benchmark in DuckDuckGo: https://blog.dubbelboer.com/2015/08/23/rwmutex-vs-atomicvalue-vs-unsafepointer.html

Reply:
I thought about it. But decided not to change the logic significantly